[Panorama Viewer] Add PanoramaHDR editing

N0Ls opened the pull request:
This PR introduces the following features :
- Create a PanoramaViewer class
- Update the FloatImageViewer class to display distortion
- Add surface class to handle discretization of panorama

This PR is linked to the following in the meshroom repository : https://github.com/alicevision/meshroom/pull/1349

Thomas-Zorroche committed:
wip [PanoramaViewer] add first version

Thomas-Zorroche committed:
[PanoramaViewer] display grid by default

Thomas-Zorroche committed:
[PanoramaViewer] Grid reload fix

Thomas-Zorroche committed:
[PanoramaViewer] grid update on changes

Thomas-Zorroche committed:
[Distortion] Add Surface class and Distortion

Thomas-Zorroche committed:
CmakeList fix : add surface files

Thomas-Zorroche committed:
[Distortion] grid follows distortion

Thomas-Zorroche committed:
[Distortion] : fix hdr viewer

Thomas-Zorroche committed:
[Distortion] : fix grid display

Thomas-Zorroche committed:
[Distortion] Dynamic subdivisions

N0Ls committed:
[Distortion] fix subdivisions update problem

Thomas-Zorroche committed:
[Distortion] fix subdivisions grid

Change indexCount and VertexCount when subs are changed

Thomas-Zorroche committed:
[Distortion] fix init distortion problem

N0Ls committed:
[Distortion] change display grid function

Thomas-Zorroche committed:
[Distortion] fix distortion issues when switching between viewers

Thomas-Zorroche committed:
[Distortion] dynamic sfm data

Thomas-Zorroche committed:
[Distortion] no disto in HDR viewer

Thomas-Zorroche committed:
[Distortion] Principal Point  + dynamic Intrinsic

Thomas-Zorroche committed:
[Distortion] refactoring FloatImageViewer header

Thomas-Zorroche committed:
[Distortion] refactoring FloatImageViewer cpp

Thomas-Zorroche committed:
[Distortion] refactoring Surface code

Thomas-Zorroche committed:
[Distortion] change subdivisions stepsize to 2

Thomas-Zorroche committed:
[Distortion] use lowerCamelCase for methods

Thomas-Zorroche committed:
[Distortion] Remove panorama viewer

Thomas-Zorroche committed:
[PanoramaViewer] add initial code

Thomas-Zorroche committed:
[PanoramaViewer] add imagesData QVariantMap

Thomas-Zorroche committed:
[Panorama Viewer] wip pano preview

Thomas-Zorroche committed:
[Panorama Viewer] fix orientation problem

Thomas-Zorroche committed:
[Panorama Viewer] remove grid inside Panorama Viewer

Thomas-Zorroche committed:
[Panorama Viewer] change Image coordinates system variables

Thomas-Zorroche committed:
[Panorama Viewer] QML Link Rotation Panorama Values

Thomas-Zorroche committed:
[Panorama Viewer] remove indentation namespace

Thomas-Zorroche committed:
[Panorama Viewer] refactoring surface code

Thomas-Zorroche committed:
[Panorama Viewer] ViewerType enum

Thomas-Zorroche committed:
[Panorama Viewer] remove LoadSfm boolean

Thomas-Zorroche committed:
[Panorama Viewer] updatePaintSurface private method

Thomas-Zorroche committed:
[Panorama Viewer] WIP Rotation Pano

Thomas-Zorroche committed:
[Panorama Viewer] Mouse Area Coords for highlights (double gamma)

Thomas-Zorroche committed:
[Panorama Viewer] downscale images for Panorama Viewer

Thomas-Zorroche committed:
[Panorama Viewer] Display Full panorama without issues

Thomas-Zorroche committed:
[Panorama Viewer] WIP Rotate Pano Real Time

Thomas-Zorroche committed:
[Panorama Viewer] Inverse gamma on hovering

Thomas-Zorroche committed:
[Panorama Viewer] Fix Rotation issue on seem

Thomas-Zorroche committed:
[Panorama Viewer] WIP fix Spherical Pano cut issue

There are still some problem in the top part of the image

Thomas-Zorroche committed:
[Panorama Viewer] add downscale panorama function

Thomas-Zorroche committed:
[Panorama Viewer] Update Subdivisions for panorama

Thomas-Zorroche committed:
[Panorama Viewer] reload FloatImageViewer when downscale changes

Thomas-Zorroche committed:
[Panorama Viewer] Panorama cpp / hpp  : code refactoring

Thomas-Zorroche committed:
[Panorama Viewer] new rotation system with yaw and pitch

Thomas-Zorroche committed:
[Panorama Viewer] fix pitch issue

Thomas-Zorroche committed:
[Panorama Viewer] fix order of transformation for pano rotation

Thomas-Zorroche committed:
[Panorama Viewer] getter for pitch and yaw

Thomas-Zorroche committed:
[Panorama Viewer] New projection method for panorama stiching

Thomas-Zorroche committed:
[Panorama Viewer] correct ratio + fix deleted column bug

Thomas-Zorroche committed:
[Panorama Viewer] Warp images correctly

Thomas-Zorroche committed:
[Panorama Viewer] Correct Pitch and Yaw Values

Thomas-Zorroche committed:
[Panorama Viewer] change pitch and yaw from SfMTransform Node

Thomas-Zorroche committed:
[Panorama Viewer] Ensure panorama fits in RAM memory before loading

Thomas-Zorroche committed:
[Panorama Viewer] domain of pitch [-90;90]

Thomas-Zorroche committed:
[Panorama Viewer] remove debug line for downscale

Thomas-Zorroche committed:
[Panorama Viewer] remove some debug lines (ram)

Thomas-Zorroche committed:
[Panorama Viewer] new algorithm for highlight images

Thomas-Zorroche committed:
[Panorama Viewer] remove rotation from float image viewer

Thomas-Zorroche committed:
[Panorama Viewer] delete unused functions for highlight

N0Ls committed:
[Panorama Viewer] Correcting principal point misplacement

Thomas-Zorroche committed:
[Panorama Viewer] change highlight values

fabiencastan commented on src/PanoramaViewer.cpp:
```diff
@@ -0,0 +1,112 @@
+        totalSizeImages *= 0.5;
+    }
+
+    _imagesData.insert("lvl", downscaleLevel);
```
Why hidding that in the list of images?
Just use a dedicated property for that.

Is there any reason to dislike vowels? Always choose names that reduce unambiguities as much as possible.

fabiencastan commented on src/Surface.cpp:
```diff
@@ -0,0 +1,397 @@
+    for (int j = 0; j < _subdivisions; j++) {
+        for (int i = 0; i < _subdivisions; i++) {
+            int remove = 0;
+            for (const auto it : _deletedColIndex) {
```
Just use an array of boolean with the "enabled" status.

fabiencastan commented on src/Surface.cpp:
```diff
@@ -0,0 +1,397 @@
+
+void Surface::computeIndicesGrid(quint16* indices)
+{
+    int pointer = 0;
```
int index = 0;

fabiencastan commented on src/FloatImageViewer.hpp:
```diff
@@ -96,18 +118,45 @@ class FloatImageViewer : public QQuickItem
     Q_INVOKABLE QVector4D pixelValueAt(int x, int y);
+    Q_INVOKABLE QPoint getVertex(int index);
+    Q_INVOKABLE void setVertex(int index, float x, float y);
+    Q_INVOKABLE void displayGrid(bool display);
+    Q_INVOKABLE void setGridColorQML(const QColor& color);
+    Q_INVOKABLE void defaultControlPoints();
+    Q_INVOKABLE void resized();
+    Q_INVOKABLE bool reinit();
+    Q_INVOKABLE void hasDistortion(bool distortion);
+    Q_INVOKABLE void updateSubdivisions(int subs);
+    Q_INVOKABLE void setSfmPath(const QString& path);
+    Q_INVOKABLE QPoint getPrincipalPoint();
+    Q_INVOKABLE void setIdView(int id);
+    Q_INVOKABLE void setPanoViewerEnabled(bool state);
```
use properties instead of invokable

fabiencastan commented on src/Surface.cpp:
```diff
@@ -0,0 +1,397 @@
+aliceVision::Vec2 toEquirectangular(const aliceVision::Vec3& spherical, int width, int height);
+
+// Static Variables Initialisation
+int Surface::_downscaleLevelPanorama = 0;
+const int Surface::_panoramaWidth = 3000;
+const int Surface::_panoramaHeight = 1500;
```
no reason to be static const
just use plain variables

fabiencastan commented on src/Surface.hpp:
```diff
@@ -0,0 +1,178 @@
+	int _subdivisions;
+	int _vertexCount;
+	int _indexCount;
+	std::vector<std::pair<int, int> > _deletedColIndex;
```
std::vector<bool> _vertexEnabled;

fabiencastan commented on src/Surface.hpp:
```diff
@@ -0,0 +1,178 @@
+	/*
+	* Static Functions
+	*/
+	static int downscaleLevelPanorama() { return _downscaleLevelPanorama; }
```
The downscale is a notion for the Image not for the Surface.

fabiencastan commented on src/FloatImageViewer.hpp:
```diff
@@ -96,18 +118,45 @@ class FloatImageViewer : public QQuickItem
+    Q_SIGNAL void sfmChanged();
 
     Q_INVOKABLE QVector4D pixelValueAt(int x, int y);
+    Q_INVOKABLE QPoint getVertex(int index);
```
These should be properties of the surface, and not of the Image.

Thomas-Zorroche committed:
Add a Q_PROPERY Surface to the FIV

Add a Q_PROPERTY Color to the Surface

N0Ls committed:
[Panorama Viewer] Change the image highlight handling

We just boost the gamma by 1.0 so it work with the edit inside the preview.

Thomas-Zorroche committed:
[Panorama Viewer] Refactor downscale attribute

Use Qt Q_Property

Thomas-Zorroche committed:
[Panorama Viewer] Rename pointer to index

N0Ls committed:
[Panorama Viewer] Move getter to the surface instead of image

N0Ls committed:
[Panorama Viewer] Move logic to surface

Moving getters for yaw and pitch and the isMouseInside function.

N0Ls committed:
[Panorama Viewer] Remove dead function reinit

Thomas-Zorroche committed:
[Panorama Viewer] refactor surface getters for vertices

Thomas-Zorroche committed:
[Panorama Viewer] refactor gridColor getters

Thomas-Zorroche committed:
[Panorama Viewer] (WIP) change displayGrid logic

Thomas-Zorroche committed:
[Panorama Viewer] Add Q_Property for MouseOver

Thomas-Zorroche committed:
[Panorama Viewer] Add Q_Property for SfmPath

Thomas-Zorroche committed:
[Panorama Viewer] (WIP) Subdivisions not working with property 

Here it's working with Q_Invokable

Thomas-Zorroche committed:
[Panorama Viewer] Move rotate methods to Surface Class

Thomas-Zorroche committed:
[Panorama Viewer] Add Q_ENUM for EViewerType

N0Ls committed:
[Lens Distortion Viewer] Change grid opacity with qProperty

We now use a qProperty to change the grid opacity so it works with the color change.

Thomas-Zorroche committed:
[Panorama Viewer] Fix downscale issue

Thomas-Zorroche committed:
[Float Image Viewer] reload sfmData when vertices are empty

N0Ls committed:
[Panorama Viewer] Move set vertices signal to surface class

Thomas-Zorroche committed:
[Panorama Viewer] Add qProperty for downscale

Thomas-Zorroche committed:
[Panorama Viewer] remove dead code

Thomas-Zorroche committed:
[Panorama Viewer] add qProperty for downscale and sfmPath for Panorama

N0Ls committed:
[Panorama Viewer] Reorganize code in surface class

N0Ls committed:
[Panorama Viewer] Clean code in float image class

N0Ls committed:
[Panorama Viewer] Rename methods and attributes for more clarity

Thomas-Zorroche committed:
[Panorama Viewer] Change fonction to retrieve principal point

Thomas-Zorroche committed:
[Panorama Viewer] add 2D Boolean vector for enable vertex on seems

Change from a vector of pair to a 2D boolean vector in order to handle images on seems and decide whether or not these vertices should be drawn or not

Thomas-Zorroche committed:
[Panorama Viewer] prevent to reload twice when changing downscale

Thomas-Zorroche committed:
[Panorama Viewer] Downscale is 4 by default

fabiencastan commented on src/FloatImageViewer.hpp:
```diff
@@ -121,12 +154,18 @@ class FloatImageViewer : public QQuickItem
+    // Prevent to update surface without the root created
+    bool _createRoot = true;
+    // Level of downscale for images of a Panorama
+    int _downscaleLevel = 4;
```
```suggestion
    int _downscaleLevel = 0;
```

And should be set to a bigger value only in the Panorama.

N0Ls committed:
[Panorama Viewer] Add HDR viewer status method

Add metod to check if we the HDR viewer is enabled. 
Its use avoid to have a display problem after changing scene.

Thomas-Zorroche committed:
[Panorama Viewer] Prevent division by 0

Thomas-Zorroche committed:
[Panorama Viewer] Fix bug on reload images

Thomas-Zorroche committed:
[Panorama Viewer] Do not change subdivisions on reload images

N0Ls committed:
[Panorama Viewer] Remove viewer type interdependency for hover

Use a boolean set in the QML instead of checking the viewer type within the float image viewer.

N0Ls committed:
[Panorama Viewer] Change downscale handling to avoid interdependency

Thomas-Zorroche committed:
[Panorama Viewer] fix vertices bug on reload disto viewer

N0Ls committed:
[Panorama Viewer] Change grid handling in disto viewer

N0Ls committed:
[Panorama Viewer] Add qProperty 'SfM required' to avoid interdependency

N0Ls committed:
[Panorama Viewer] Remove viewer type reference within float image

Thomas-Zorroche committed:
[Panorama Viewer] add mSfmData property and new logic for compute grid

Thomas-Zorroche committed:
[Panorama Viewer] Update surface only if image is not null

Thomas-Zorroche committed:
[Panorama Viewer] compute with intrinsics for subdivisions update

Thomas-Zorroche committed:
[Panorama Viewer] computeVerticesGrid: fix intrinsic bug

Thomas-Zorroche committed:
[Panorama Viewer] Clear vertices onSfmUpdate

Thomas-Zorroche committed:
[Panorama Viewer] remove texture size from ppPoint argument function

Thomas-Zorroche committed:
[Panorama Viewer] rename some variable

Thomas-Zorroche committed:
[Panorama Viewer] Surface: clean up methods code

N0Ls committed:
[Panorama Viewer] Limit pitch angle between 180° and -180°

Allows to access all possible angles

N0Ls committed:
[Panorama Viewer] Add roll modification preview

Thomas-Zorroche committed:
[Panorama Viewer] SetRotation: Limit pitch value between [-180, 180]

N0Ls committed:
[Panorama Viewer] Add roll getter

Thomas-Zorroche committed:
[Panorama Viewer] clean up euler angles getters code

Thomas-Zorroche committed:
[Panorama Viewer] rename surface rotation methods

N0Ls committed:
[Panorama Viewer] Add roll increment to euler angles incrementation

N0Ls committed:
[MSfMData] Add event to camera number

N0Ls committed:
[wip] Use qt binding on angles for edition

N0Ls committed:
[wip] Add qProperties for pitch and roll

Thomas-Zorroche committed:
[Panorama Viewer] remove unnecessary rotate functions

Thomas-Zorroche committed:
[Panorama Viewer] Add msfmData

N0Ls committed:
[mSfMData] Expose views ids

N0Ls committed:
[mSfMData] Change return type from QUrl to Qstring

N0Ls committed:
[wip] Add outdated status to mSfMData

Thomas-Zorroche committed:
[wip] fix outdated issues MSfmData

Thomas-Zorroche committed:
[Panorama Viewer] connect msfmData to surface signal

N0Ls committed:
[wip] Connect msfdata signal to panorama

Thomas-Zorroche committed:
[Panorama Viewer] disable Q_EMIT from setMSfmData Pano

N0Ls committed:
[Panorama Viewer] Remove debugs

N0Ls committed:
[Panorama Viewer] Remove sfmLoaded bool because of duplicate use

N0Ls committed:
[Panorama Viewer] Add fisheye support for viewer

Thomas-Zorroche committed:
[Panorama Viewer] fix MSfmData null pointer case

Thomas-Zorroche committed:
[Surface] fix isMouseInside function

Go to next triangle instead of next image

N0Ls committed:
[Disto Viewer] Refactor code and compute principal point on the fly

N0Ls committed:
[Panorama Viewer] Change the method to get the fisheye circle parameters

N0Ls committed:
[Panorama Viewer] Add comments

fabiencastan committed:
[cmake] direct dependency to ilmbase not required

Tested on linux
